service/src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java

```diff
@@ -312,60 +314,35 @@ public void deleteGcpCloudContext(UUID workspaceId, AuthenticatedUserRequest use
         .submitAndWait(null);
   }
 
-  /**
-   * Retrieve the optional GCP cloud context NOTE: no user auth validation
-   *
-   * @param workspaceId workspace identifier of the cloud context
-   * @return optional GCP cloud context
-   */
-  public Optional<GcpCloudContext> getGcpCloudContext(UUID workspaceId) {
-    return workspaceDao.getGcpCloudContext(workspaceId);
-  }
-
-  /**
-   * Retrieve the optional GCP cloud context, providing a workspace This is a frequent usage, so we
-   * make a method for it to save coding the fetch of workspace id every time NOTE: no user auth
-   * validation
-   */
-  public Optional<GcpCloudContext> getGcpCloudContext(Workspace workspace) {
-    return workspaceDao.getGcpCloudContext(workspace.getWorkspaceId());
-  }
-
-  /**
-   * Helper method for looking up the GCP project ID for a given workspace ID, if one exists. Unlike
-   * {@link #getRequiredGcpProject(UUID)}, this returns an empty Optional instead of throwing if the
-   * given workspace does not have a GCP cloud context. NOTE: no user auth validation
-   *
-   * @param workspaceId workspace identifier of the cloud context
-   * @return optional GCP project from the cloud context
-   */
-  public Optional<String> getGcpProject(UUID workspaceId) {
-    return workspaceDao.getGcpCloudContext(workspaceId).map(GcpCloudContext::getGcpProjectId);
+  /** Get an optional cloud context */
+  public Optional<GcpCloudContext> getGcpCloudContext(
+      UUID workspaceId, AuthenticatedUserRequest userRequest) {
+    validateWorkspaceAndAction(userRequest, workspaceId, SamConstants.SAM_WORKSPACE_READ_ACTION);
+    return gcpCloudContextService.getGcpCloudContext(workspaceId);
   }
 
   /**
-   * Helper method used by other classes that require the GCP project to exist in the workspace. It
-   * throws if the project (GCP cloud context) is not set up. NOTE: no user auth validation
+   * Get an option project id from an optional GCP cloud context
    *
    * @param workspaceId unique workspace id
-   * @return GCP project id
+   * @param userRequest authenticated user
+   * @return optional project id
    */
-  public String getRequiredGcpProject(UUID workspaceId) {
-    return getGcpProject(workspaceId)
-        .orElseThrow(
-            () -> new CloudContextRequiredException("Operation requires GCP cloud context"));
+  public Optional<String> getGcpProject(UUID workspaceId, AuthenticatedUserRequest userRequest) {
+    validateWorkspaceAndAction(userRequest, workspaceId, SamConstants.SAM_WORKSPACE_READ_ACTION);
+    return gcpCloudContextService.getGcpProject(workspaceId);
   }
 
   /**
-   * Get the GcpProject with authentication for use in controllers
+   * Get the GcpProject or throw
    *
    * @param workspaceId unique workspace id
    * @param userRequest authenticated user
    * @return GCP project id
    */
   public String getRequiredGcpProject(UUID workspaceId, AuthenticatedUserRequest userRequest) {
```

Comment:
I'd implement this in terms of the other one: `return getGcpProject(workspaceId, userRequest).orElseThrow(() -> ...)`

Reply:
That would put the error logic in two places: in GcpCloudContextService and here. I think it is better as is.